src/System.Private.ServiceModel/src/System/ServiceModel/UpnEndpointIdentity.cs

```diff
@@ -53,6 +57,110 @@ internal UpnEndpointIdentity(WindowsIdentity windowsIdentity)
             _upnSid = windowsIdentity.User;
             _hasUpnSidBeenComputed = true;
         }
+
+        internal override void EnsureIdentityClaim()
+        {
+            if (_windowsIdentity != null)
+            {
+                lock (_thisLock)
+                {
+                    if (_windowsIdentity != null)
+                    {
+                        Initialize(Claim.CreateUpnClaim(GetUpnFromWindowsIdentity(_windowsIdentity)));
+                        _windowsIdentity.Dispose();
+                        _windowsIdentity = null;
+                    }
+                }
+            }
+        }
+
+        string GetUpnFromWindowsIdentity(WindowsIdentity windowsIdentity)
+        {
+            string downlevelName = null;
+            string upnName = null;
+ 
+            try
+            {
+                downlevelName = windowsIdentity.Name;
+ 
+                if (IsMachineJoinedToDomain())
```

Comment:
I was trying to keep the code as identical as possible only throwing when we get to some code we can implement. Looking at the code, downlevelName is in the format domain\username. The fully qualified domain name is username@fqdn.tld. I'll give this some more thought. We do need to find a solution to this but none of the code in corefx uses the P/Invoke api call that we use on full framework so we can't expect to find an alternative way to do this using framework api's, at least not yet.

---
In reply to: [77633040](https://github.com/dotnet/wcf/pull/1502#discussion_r77633040) [](ancestors = 77633040,77481060)